### PR TITLE
qt 6.2.1

### PR DESCRIPTION
**Pull request by yurikoles**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by viking1304** [comment]
> Is there a similar patch for Qt5? I think it needs a similar one. Or will the same patch apply to Qt5?

@carlocab for qt5 you can use this patch that I just made

```
--- a/qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h
+++ b/qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h
@@ -43,4 +43,6 @@
 #include <qpa/qplatformgraphicsbuffer.h>
 #include <private/qcore_mac_p.h>
+ 
+#include <CoreGraphics/CGColorSpace.h>

 QT_BEGIN_NAMESPACE
```

Both qt5 and qt6 requires exactly the same change in `qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h` but original patch will be rejected on qt5, because that file in qt5 doesn't have `#include <IOSurface/IOSurface.h>` line that is referenced in the patch.

**Comment by viking1304** [comment]
I managed to compile Qt 6.2.1 with `CommandLineTools` instead of XCode. I had to make 2 additional patches for `qtwebengine` to accomplish that, because `chromium` can only be build if XCode is installed and active.

First problem is that `sdk_info.py` use `xcodebuild` to get version, so I made a patch that will check if CommandLineTools is selected and it will use `pkgutil --pkg-info` instead if it is.

Second patch will fix `find_sdk.py` and will use differed paths if CTL is selected that if XCode is selected.

I have no idea if it is possible to avoid use of those scripts if some parameters are provided somewhere else, so I will just provide all the information and leave to someone else with more experience in brew formulas to finish this task.

Of course I am here if you need any more information.

`sdk_info.py`  patch

```
--- a/qtwebengine/src/3rdparty/chromium/build/config/mac/sdk_info.py
+++ b/qtwebengine/src/3rdparty/chromium/build/config/mac/sdk_info.py
@@ -68,11 +68,18 @@
     settings['xcode_build'] = version_plist['ProductBuildVersion']
     return
 
-  lines = subprocess.check_output(['xcodebuild',
-                                   '-version']).decode('UTF-8').splitlines()
-  settings['xcode_version'] = FormatVersion(lines[0].split()[-1])
+  if 'CommandLineTools' in subprocess.check_output(['xcode-select', '-p']).decode('UTF-8'):  
+    lines = subprocess.check_output(['pkgutil',
+                                     '--pkg-info=com.apple.pkg.CLTools_Executables']).decode('UTF-8').splitlines()
+    settings['xcode_version'] = FormatVersion(lines[1].split()[-1]) 
+    settings['xcode_build'] = lines[1].split(".")[-1]
+  else:
+    lines = subprocess.check_output(['xcodebuild',
+                                     '-version']).decode('UTF-8').splitlines()
+    settings['xcode_version'] = FormatVersion(lines[0].split()[-1])
+    settings['xcode_build'] = lines[-1].split()[-1]
+
   settings['xcode_version_int'] = int(settings['xcode_version'], 10)
-  settings['xcode_build'] = lines[-1].split()[-1]
```

`find_sdk.py` patch

```
--- a/qtwebengine/src/3rdparty/chromium/build/mac/find_sdk.py
+++ b/qtwebengine/src/3rdparty/chromium/build/mac/find_sdk.py
@@ -66,12 +66,16 @@
     print(err, file=sys.stderr)
     raise Exception('Error %d running xcode-select' % job.returncode)
   dev_dir = out.decode('UTF-8').rstrip()
-  sdk_dir = os.path.join(
-      dev_dir, 'Platforms/MacOSX.platform/Developer/SDKs')
+  if 'CommandLineTools' in dev_dir:
+    sdk_dir = os.path.join(
+        dev_dir, 'SDKs')
+  else:
+    sdk_dir = os.path.join(
+        dev_dir, 'Platforms/MacOSX.platform/Developer/SDKs')

   if not os.path.isdir(sdk_dir):
     raise SdkError('Install Xcode, launch it, accept the license ' +
-      'agreement, and run `sudo xcode-select -s /path/to/Xcode.app` ' +
+      'agreement, an<d run `sudo xcode-select -s /path/to/Xcode.app` ' +
       'to continue.')
   sdks = [re.findall('^MacOSX(\d+\.\d+)\.sdk$', s) for s in os.listdir(sdk_dir)]
   sdks = [s[0] for s in sdks if s]  # [['10.5'], ['10.6']] => ['10.5', '10.6']
@@ -86,7 +90,10 @@
     print(os.path.join(sdk_dir, sdk_name))

   if options.print_bin_path:
-    bin_path = 'Toolchains/XcodeDefault.xctoolchain/usr/bin/'
+    if 'CommandLineTools' in dev_dir:
+      bin_path = 'usr/bin/'
+    else:
+      bin_path = 'Toolchains/XcodeDefault.xctoolchain/usr/bin/'
     print(os.path.join(dev_dir, bin_path))

   return best_sdk
```

**Comment by SMillerDev** [comment]
Could you try upstreaming these patches to chromium? I bet they're useful to more than just QT in homebrew.

**Comment by viking1304** [comment]
> Could you try upstreaming these patches to chromium? I bet they're useful to more than just QT in homebrew.

When I do something I really want to understand what I am doing and why, but I have no idea if they intentionally require XCode for some reason or this is just a simple oversight.

In order to submit patches upstream both `sdk_info.py` and `find_sdk.py` should be completely patched, since partial patches might cause additional issues later. Upstream patch should at least also fix `FillSDKPathAndVersion` in `sdk_info.py` and probably few more things.

For example when you use my patch with`sdk_info.py` this will work

```
sdk_info.py macosx
```

but you will still get an error if you run this

```
sdk_info.py --get_sdk_info macosx
```

Those patches that I made are just workaround patches to allow build of qt 6.2.1 with CTL, which is still better that to skip `qtwebengine` completely (which is done in qt5 formula).

**Comment by danielnachun** [comment]
How important is it for us to be able to build this with the CLT instead of Xcode?  This is a relocatable bottle so nobody other than a few maintainers and our CI system have to build it from source.  That's not to say we shouldn't provide fixes to upstream projects when we find them incidentally, but in this case if it's a lot of work to upstream the necessary fixes and we can just add Xcode as a build dependency, maybe the additional build dependency is the easier path here.

**Comment by yurikoles** [comment]
Could you please add label `CI-no-fail-fast`?

**Comment by SMillerDev** [comment]
The error is because of a 404 from the QT source, I don't think it will start working with `CI-no-fail-fast`

**Comment by carlocab** [comment]
Adding an Xcode build dependency might seem easier, but this could also create an implicit runtime dependency on a full Xcode installation. We likely won't be able to detect this in CI because we don't test on CLT-only systems.

**Comment by yurikoles** [comment]
I'm going to build it locally and run `linkage`.

**Comment by viking1304** [comment]
> How important is it for us to be able to build this with the CLT instead of Xcode? This is a relocatable bottle so nobody other than a few maintainers and our CI system have to build it from source. That's not to say we shouldn't provide fixes to upstream projects when we find them incidentally, but in this case if it's a lot of work to upstream the necessary fixes and we can just add Xcode as a build dependency, maybe the additional build dependency is the easier path here.

Let me try to explain what exactly happening and why, so you guys can better understand the problem and hopefully find the best solution. 

There are two main problems here:

1. `qiosurfacegraphicsbuffer.h` needs a missing header ([QTBUG-97667](https://bugreports.qt.io/browse/QTBUG-97667)). Solution is to use this [patch](https://code.qt.io/cgit/qt/qtbase.git/patch/src/plugins/platforms/cocoa?id=dece6f5840463ae2ddf927d65eb1b3680e34a547) and there is no alternatives.

2. `qtwebengine` package cannot be build without XCode, since it requires `xcodebuild` and paths that exist only in XCode and do not exist in CTL. I tried to use `depends_on :xcode` but it still failed. That's why I made those patches.

So, you have few choices:
a) to skip `qtwebengine` as you did in qt5 formula
b) to use my patches in order to be able to compile `qtwebengine` in which case you don't need xcode depnedancy
c) to find some other way which I can't see at this moment

> The error is because of a 404 from the QT source

This is strange, both url and sha256 are correct and link works. Just checked.

btw. what is the difference between `depends_on :xcode` and `depends_on xcode:`? I can't find that info anywhere, but I see that qt6 use `xcode:` and not `:xcode` which is in documentation.

**Comment by yurikoles** [comment]
You need to take into account full string: `depends_on xcode: [:build]`, here we supply parameter after colon.

**Comment by yurikoles** [comment]
Here is the linkage output for build made with Xcode 13.1 @ Monterey 12.1 @ M1:

<details>
	<summary>brew linkage qt</summary>

```
System libraries:
  /usr/lib/libSystem.B.dylib
  /usr/lib/libbsm.0.dylib
  /usr/lib/libc++.1.dylib
  /usr/lib/libcups.2.dylib
  /usr/lib/libobjc.A.dylib
  /usr/lib/libpmenergy.dylib
  /usr/lib/libpmsample.dylib
  /usr/lib/libresolv.9.dylib
  /usr/lib/libsandbox.1.dylib
  /usr/lib/libz.1.dylib
Homebrew libraries:
  /opt/homebrew/opt/assimp/lib/libassimp.5.dylib (assimp)
  /opt/homebrew/opt/brotli/lib/libbrotlidec.1.dylib (brotli)
  /opt/homebrew/opt/dbus/lib/libdbus-1.3.dylib (dbus)
  /opt/homebrew/opt/double-conversion/lib/libdouble-conversion.3.dylib (double-conversion)
  /opt/homebrew/opt/freetype/lib/libfreetype.6.dylib (freetype)
  /opt/homebrew/opt/glib/lib/libglib-2.0.0.dylib (glib)
  /opt/homebrew/opt/glib/lib/libgthread-2.0.0.dylib (glib)
  /opt/homebrew/opt/hunspell/lib/libhunspell-1.7.0.dylib (hunspell)
  /opt/homebrew/opt/icu4c/lib/libicudata.69.dylib (icu4c)
  /opt/homebrew/opt/icu4c/lib/libicui18n.69.dylib (icu4c)
  /opt/homebrew/opt/icu4c/lib/libicuuc.69.dylib (icu4c)
  /opt/homebrew/opt/jasper/lib/libjasper.4.dylib (jasper)
  /opt/homebrew/opt/jpeg/lib/libjpeg.9.dylib (jpeg)
  /opt/homebrew/opt/libb2/lib/libb2.1.dylib (libb2)
  /opt/homebrew/opt/libpng/lib/libpng16.16.dylib (libpng)
  /opt/homebrew/opt/libtiff/lib/libtiff.5.dylib (libtiff)
  /opt/homebrew/opt/libxcb/lib/libxcb-render.0.dylib (libxcb)
  /opt/homebrew/opt/libxcb/lib/libxcb-shape.0.dylib (libxcb)
  /opt/homebrew/opt/libxcb/lib/libxcb-xfixes.0.dylib (libxcb)
  /opt/homebrew/opt/libxcb/lib/libxcb.1.dylib (libxcb)
  /opt/homebrew/opt/md4c/lib/libmd4c.0.dylib (md4c)
  /opt/homebrew/opt/openssl@1.1/lib/libcrypto.1.1.dylib (openssl@1.1)
  /opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib (openssl@1.1)
  /opt/homebrew/opt/pcre2/lib/libpcre2-16.0.dylib (pcre2)
  /opt/homebrew/opt/sqlite/lib/libsqlite3.0.dylib (sqlite)
  /opt/homebrew/opt/webp/lib/libwebp.7.dylib (webp)
  /opt/homebrew/opt/webp/lib/libwebpdemux.2.dylib (webp)
  /opt/homebrew/opt/webp/lib/libwebpmux.3.dylib (webp)
Indirect dependencies with linkage:
  openssl@1.1
Variable-referenced libraries:
  @rpath/Qt3DAnimation.framework/Versions/A/Qt3DAnimation
  @rpath/Qt3DCore.framework/Versions/A/Qt3DCore
  @rpath/Qt3DExtras.framework/Versions/A/Qt3DExtras
  @rpath/Qt3DInput.framework/Versions/A/Qt3DInput
  @rpath/Qt3DLogic.framework/Versions/A/Qt3DLogic
  @rpath/Qt3DQuick.framework/Versions/A/Qt3DQuick
  @rpath/Qt3DQuickAnimation.framework/Versions/A/Qt3DQuickAnimation
  @rpath/Qt3DQuickExtras.framework/Versions/A/Qt3DQuickExtras
  @rpath/Qt3DQuickInput.framework/Versions/A/Qt3DQuickInput
  @rpath/Qt3DQuickRender.framework/Versions/A/Qt3DQuickRender
  @rpath/Qt3DQuickScene2D.framework/Versions/A/Qt3DQuickScene2D
  @rpath/Qt3DRender.framework/Versions/A/Qt3DRender
  @rpath/QtBodymovin.framework/Versions/A/QtBodymovin
  @rpath/QtCharts.framework/Versions/A/QtCharts
  @rpath/QtChartsQml.framework/Versions/A/QtChartsQml
  @rpath/QtConcurrent.framework/Versions/A/QtConcurrent
  @rpath/QtCore.framework/Versions/A/QtCore
  @rpath/QtDBus.framework/Versions/A/QtDBus
  @rpath/QtDataVisualization.framework/Versions/A/QtDataVisualization
  @rpath/QtDesigner.framework/Versions/A/QtDesigner
  @rpath/QtDesignerComponents.framework/Versions/A/QtDesignerComponents
  @rpath/QtGui.framework/Versions/A/QtGui
  @rpath/QtHelp.framework/Versions/A/QtHelp
  @rpath/QtHunspellInputMethod.framework/Versions/A/QtHunspellInputMethod
  @rpath/QtLabsAnimation.framework/Versions/A/QtLabsAnimation
  @rpath/QtLabsFolderListModel.framework/Versions/A/QtLabsFolderListModel
  @rpath/QtLabsQmlModels.framework/Versions/A/QtLabsQmlModels
  @rpath/QtLabsSettings.framework/Versions/A/QtLabsSettings
  @rpath/QtLabsSharedImage.framework/Versions/A/QtLabsSharedImage
  @rpath/QtLabsWavefrontMesh.framework/Versions/A/QtLabsWavefrontMesh
  @rpath/QtMultimedia.framework/Versions/A/QtMultimedia
  @rpath/QtMultimediaQuick.framework/Versions/A/QtMultimediaQuick
  @rpath/QtNetwork.framework/Versions/A/QtNetwork
  @rpath/QtOpcUa.framework/Versions/A/QtOpcUa
  @rpath/QtOpenGL.framework/Versions/A/QtOpenGL
  @rpath/QtOpenGLWidgets.framework/Versions/A/QtOpenGLWidgets
  @rpath/QtPositioning.framework/Versions/A/QtPositioning
  @rpath/QtPositioningQuick.framework/Versions/A/QtPositioningQuick
  @rpath/QtPrintSupport.framework/Versions/A/QtPrintSupport
  @rpath/QtQml.framework/Versions/A/QtQml
  @rpath/QtQmlCore.framework/Versions/A/QtQmlCore
  @rpath/QtQmlLocalStorage.framework/Versions/A/QtQmlLocalStorage
  @rpath/QtQmlModels.framework/Versions/A/QtQmlModels
  @rpath/QtQmlWorkerScript.framework/Versions/A/QtQmlWorkerScript
  @rpath/QtQmlXmlListModel.framework/Versions/A/QtQmlXmlListModel
  @rpath/QtQuick.framework/Versions/A/QtQuick
  @rpath/QtQuick3D.framework/Versions/A/QtQuick3D
  @rpath/QtQuick3DAssetImport.framework/Versions/A/QtQuick3DAssetImport
  @rpath/QtQuick3DAssetUtils.framework/Versions/A/QtQuick3DAssetUtils
  @rpath/QtQuick3DEffects.framework/Versions/A/QtQuick3DEffects
  @rpath/QtQuick3DHelpers.framework/Versions/A/QtQuick3DHelpers
  @rpath/QtQuick3DIblBaker.framework/Versions/A/QtQuick3DIblBaker
  @rpath/QtQuick3DParticles.framework/Versions/A/QtQuick3DParticles
  @rpath/QtQuick3DRuntimeRender.framework/Versions/A/QtQuick3DRuntimeRender
  @rpath/QtQuick3DUtils.framework/Versions/A/QtQuick3DUtils
  @rpath/QtQuickControls2.framework/Versions/A/QtQuickControls2
  @rpath/QtQuickControls2Impl.framework/Versions/A/QtQuickControls2Impl
  @rpath/QtQuickDialogs2.framework/Versions/A/QtQuickDialogs2
  @rpath/QtQuickDialogs2QuickImpl.framework/Versions/A/QtQuickDialogs2QuickImpl
  @rpath/QtQuickDialogs2Utils.framework/Versions/A/QtQuickDialogs2Utils
  @rpath/QtQuickLayouts.framework/Versions/A/QtQuickLayouts
  @rpath/QtQuickParticles.framework/Versions/A/QtQuickParticles
  @rpath/QtQuickShapes.framework/Versions/A/QtQuickShapes
  @rpath/QtQuickTemplates2.framework/Versions/A/QtQuickTemplates2
  @rpath/QtQuickTest.framework/Versions/A/QtQuickTest
  @rpath/QtQuickTimeline.framework/Versions/A/QtQuickTimeline
  @rpath/QtQuickWidgets.framework/Versions/A/QtQuickWidgets
  @rpath/QtRemoteObjects.framework/Versions/A/QtRemoteObjects
  @rpath/QtRemoteObjectsQml.framework/Versions/A/QtRemoteObjectsQml
  @rpath/QtScxml.framework/Versions/A/QtScxml
  @rpath/QtScxmlQml.framework/Versions/A/QtScxmlQml
  @rpath/QtSensors.framework/Versions/A/QtSensors
  @rpath/QtSensorsQuick.framework/Versions/A/QtSensorsQuick
  @rpath/QtSerialBus.framework/Versions/A/QtSerialBus
  @rpath/QtSerialPort.framework/Versions/A/QtSerialPort
  @rpath/QtShaderTools.framework/Versions/A/QtShaderTools
  @rpath/QtSql.framework/Versions/A/QtSql
  @rpath/QtStateMachine.framework/Versions/A/QtStateMachine
  @rpath/QtStateMachineQml.framework/Versions/A/QtStateMachineQml
  @rpath/QtSvg.framework/Versions/A/QtSvg
  @rpath/QtTest.framework/Versions/A/QtTest
  @rpath/QtUiTools.framework/Versions/A/QtUiTools
  @rpath/QtVirtualKeyboard.framework/Versions/A/QtVirtualKeyboard
  @rpath/QtWebChannel.framework/Versions/A/QtWebChannel
  @rpath/QtWebEngineCore.framework/Versions/A/QtWebEngineCore
  @rpath/QtWebEngineQuick.framework/Versions/A/QtWebEngineQuick
  @rpath/QtWebEngineQuickDelegatesQml.framework/Versions/A/QtWebEngineQuickDelegatesQml
  @rpath/QtWebEngineWidgets.framework/Versions/A/QtWebEngineWidgets
  @rpath/QtWebSockets.framework/Versions/A/QtWebSockets
  @rpath/QtWebView.framework/Versions/A/QtWebView
  @rpath/QtWebViewQuick.framework/Versions/A/QtWebViewQuick
  @rpath/QtWidgets.framework/Versions/A/QtWidgets
  @rpath/QtXml.framework/Versions/A/QtXml
Undeclared dependencies with linkage:
  libxcb
```
</details>

**Comment by yurikoles** [comment]
Bump?

**Review by SMillerDev — DISMISSED** [review]

**Comment by danielnachun** [comment]
Before merging this, can we try to add the patch linked by @jonaski: https://codereview.qt-project.org/gitweb?p=qt/qtbase.git;a=patch;h=549a8f80a99280cf95631919c24a199415bef2d4?  This sounds like a serious regression and the patch is small.

File: Formula/qt.rb
Comment (danielnachun) [inline Formula/qt.rb:101]: Should we add an additional comment here reminding us to try to remove this patch in this next release assuming the relevant commits are part of that release.   Presumably if the next release has the commits merged, the patch will fail to apply, but it might be useful to put a remind here anyway.

Comment (yurikoles) [reply Formula/qt.rb:101]: Thanks, @danielnachun, this is good idea. But this will re-trigger a CI for a couple of hours :(

Comment (danielnachun) [reply Formula/qt.rb:101]: If we didn't need to add that other patch we could just make a CI-syntax only follow up.  But since I think that patch is important we can probably add the comment while we're at it.

File: Formula/qt.rb
Comment (carlocab) [inline Formula/qt.rb:63]: We don't need to wrap this in an `on_macos` block, or test for `:mojave`. Just make it an unconditional
```ruby
depends_on xcode: :build
```
in the line you deleted above.

Comment (yurikoles) [reply Formula/qt.rb:63]: > Before merging this, can we try to add the patch linked by @jonaski: https://codereview.qt-project.org/gitweb?p=qt/qtbase.git;a=patch;h=549a8f80a99280cf95631919c24a199415bef2d4? This sounds like a serious regression and the patch is small.

Ok, so we will need additional CI run. @carlocab what do you need about this patch?

Comment (carlocab) [reply Formula/qt.rb:63]: Looks fine.

File: Formula/qt.rb
Comment (carlocab) [inline Formula/qt.rb:95]: The comments you added for the patches should be here. Putting it where you did makes it much easier to miss.

**Review by carlocab — APPROVED** [review]

**Comment by yurikoles** [comment]
From Linux runner:
```fatal error: cannot write PCH file: No space left on device```

**Comment by yurikoles** [comment]
Linux run is more than 9 hrs now.

**Comment by yurikoles** [comment]
Whoops, it's Big Sur @ Intel

**Comment by carlocab** [comment]
I'm away from my computer at the moment. Could someone do this when this PR is merged:
```
brew dispatch-build-bottle --upload --issue=87708 --macos=12,12-arm64 --timeout=1440 qt
```
Thanks!

**Pull request closed without merging**
